### ISSUE TITLE
Torchscript Graph: Support dynamic axes

### ISSUE THREAD

**Issue by justinchuby** [microsoft/onnx-script#390]
Build support for dynamic axes in inputs

**Comment by titaiwangms** [comment]
TODO: Try make inputs dynamic as default, and static should be specified if it's necessary.

**Comment by titaiwangms** [comment]
Tred setting all sizes dynamic as default, but it breaks easily in ORT.

The modified in graph_building.py is easy, we can simply utilize setType API when addin input

```python
torchscript_input_value = torch._C.TensorType.create_from_tensor(input_value)
rank = torchscript_input_value.dim()
torch_value.setType(torchscript_input_value.with_sizes([None] * rank))
```

It works ideally when there is no "restricted-shape nodes" involved.

However, with following model, it breaks in unexpected Expand nodes that decomposed module generated during make_fx,
```python
    def test_dynamic_axes_on_matmul(self):
        class DynamicMatMul(torch.nn.Module):
            def forward(self, x, y):
                return torch.ops.aten.matmul(x, y)

        x = torch.randn(2, 3, 6)
        y = torch.randn(2, 6, 4)
        input_x = torch.randn(2, 3, 4)
        input_y = torch.randn(2, 4, 4)
        # self.run_test(
        #     DynamicMatMul(),
        #     (x, y),
        #     input_names=["x", "y"],
        #     dynamic_axes={"x": {2: "length"}, "y": {1: "length"}},
        #     additional_test_inputs=[(input_x, input_y)],
        # )
        _run_test_with_fx_to_onnx_exporter_reference_runtime(
            DynamicMatMul(), (x, y), additional_test_inputs=[(input_x, input_y)]
        )
```
with error message:
```
onnxruntime.capi.onnxruntime_pybind11_state.InvalidArgument: [ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Non-zero status code returned while running Expand node. Name:'_0x8a230f0_n1' Status Message: invalid expand shape
```

I figure it'd be safe we do it the old way (set static as default), so dynamic_axes should be added for API.

cc @BowenBao if you have a good idea on implementing dynamic axes n new exporter.

**Comment by BowenBao** [comment]
> However, with following model, it breaks in unexpected Expand nodes that decomposed module generated during make_fx,

Does it pass onnx checker full check and onnx reference evaluator?

For shape/type in general, I'd encourage carrying over what fx graph had, and validating & amending missing ones by onnx shape/type inference.

If dynamo export w/ symbolic tracing mode works well, we should get a fx graph full of shapes and dtypes, with symbolic dimension sizes set on a as needed basis. afaik it is supported via sympy. @jcwchen please offer insight on onnx side how we could utilize that info with onnx shape inference.

If this works well we can leave `dynamic axes` in the past.

**Comment by jcwchen** [comment]
I don't have fancy tips for deciding whether to use symbolic dimension or not. I think it's really operator-by-operator. Take Expand as an example [here](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/cpu/tensor/expand.cc#L63): it cannot accept a symbolic input shape during Runtime, although onnx shape_inference is fine with it (won't throw an error for this symbolic case).

**Comment by gramalingam** [comment]
Sorry, I don't understand the issue. We can distinguish between symbolic shapes used in static-shape-inference and shapes specified as real-inputs to certain ops like Expand or Reshape. The two are not the same. I guess the issue is in the exporter when running in a mode like trace-mode (or whatever it is called, I am not sure about the name used in the exporter) about how to generalize for inputs of other shapes?

**Comment by gramalingam** [comment]
It seems to me the problems would arise when a pytorch model invokes Expand, and we don't have a graph-representation of the computation for the second input of Expand (namely, the target output shape), but we have only a constant tensor. Is that the concern/issue?

**Comment by titaiwangms** [comment]
Ahhh, I get it now. We messed up the Expand conversion. I see we use constant 1 in size to represent unchange dim, maybe that's the key.

**Comment by BowenBao** [comment]
@titaiwangms could you post the "matmul w/ expand" onnx graph for illustration?

I agree with Rama's theory. Maybe this particular issue is due to statically constructing the symbols as a constant tensor.

**Comment by titaiwangms** [comment]
![expand_matmul](https://user-images.githubusercontent.com/18010845/221731904-6f079490-8393-40a5-b52a-5ef0a95ca2f3.png)

Yes, @jcwchen just helped me figure that out. So even though I set all input sizes dynamic in torchscript, the Expand node we have right now does not take dynamic shape into consideration like we used to in old exporter:
https://github.com/microsoft/onnx-script/blob/925e81cc1736c7766e336cc470f23be432929aca/onnxscript/function_libs/torch_aten/ops/core.py#L2037-L2042

What we had
https://github.com/pytorch/pytorch/blob/b818b3fe1c1fa90529b8500cd8ef800bec8415e8/torch/onnx/symbolic_opset9.py#L909-L927

**Comment by titaiwangms** [comment]
> It seems to me the problems would arise when a pytorch model invokes Expand, and we don't have a graph-representation of the computation for the second input of Expand (namely, the target output shape), but we have only a constant tensor. Is that the concern/issue?

The expand node is auto added into fx graph (torch IR) when torch.matmul is used, and a constant size would be traced and passed to onnx::Expand, exactly like you said. I thought maybe adding -1 dim support in aten_expand would help, but the constant input size is actually the main problem. Is there anyway we can workaround it?

**Comment by gramalingam** [comment]
Didn't understand that. Changing -1 (used in aten::expand) to 1 (used in onnx::expand) can be done, I suppose, either as in the old code, or by using the `Abs` op. Are you saying the issue is some value other than -1 ?

**Comment by titaiwangms** [comment]
So with that MatMul model, torch intends to add on Expand nodes, which traces constant sizes, and it makes whole model being static. I am wondering if there is a way to workaround it. I guess it's more of a converter issue of patching this.

**Comment by BowenBao** [comment]
Sounds like fx graph problem. The expand size input was constant so there is nothing you can do. I'm afraid we can't do the old exporter style of setting dynamic axes and propagate, due to shapes being folded as constant by dynamo/fx aggressive optimization.

Hence it's back to https://github.com/microsoft/onnx-script/issues/390#issuecomment-1447229302, that we need fx to give us shape filled with symbols, and we need exporter/onnx to be able to process it. You can try altering  https://github.com/pytorch/pytorch/blob/3fa939625beed4f4b661829323e1b47b05d8f59e/torch/onnx/_internal/fx/exporter.py#L630
to `tracing_mode="symbolic",`  and comment out `decomposed_module = _shape_inference_with_fake_tensor(decomposed_module, *args)` (which does constant shape inference) and experiment.
